File: Natixis Craftsmanship Practices/Breaking Dependencies/ShoppingTunnel/cpp/src/CheckoutProcessor.cpp
#include "CheckoutProcessor.h"

#include <memory>

#include "DiscountCalculator.h"
#include "ExternalTax.h"
#include "VipDiscountCalculator.h"

int CheckoutProcessor::process(const CartData &data) {
    int discount = 0;

    std::unique_ptr<DiscountCalculator> calc;

    if (data.vip) {
        calc = std::make_unique<VipDiscountCalculator>();
    } else {
        calc = std::make_unique<DiscountCalculator>();
    }

    if (data.total > 0) {
        discount = calc->compute(data.total);
    }

    ExternalTax tax;
    int taxed = data.total - discount;

    if (taxed > 0) {
        taxed += taxed * tax.rate(data.country) / 100;
    }

    return taxed;
}

File: Natixis Craftsmanship Practices/Breaking Dependencies/ShoppingTunnel/csharp/CheckoutProcessor.cs
using ShoppingTunnel.ExternalLibrary;

namespace ShoppingTunnel;

public class CheckoutProcessor
{
    public int Process(CartData data)
    {
        var discount = 0;

        DiscountCalculator calc;
        
        if (data.Vip)
        {
            calc = new VipDiscountCalculator();
        }
        else
        {
            calc = new DiscountCalculator();
        }

        if (data.Total > 0)
        {
            discount = calc.Compute(data.Total);
        }

        var tax = new ExternalTax();
        
        var taxed = data.Total - discount;

        if (taxed > 0)
        {
            taxed += taxed * tax.Rate(data.Country) / 100;
        }

        return taxed;
    }
}